features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/domain/ResourceRequestBodyFactory.kt

```diff
@@ -0,0 +1,164 @@
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.domain
+
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.api.storage.RawBatchEvent
+import com.datadog.android.sessionreplay.internal.exception.InvalidPayloadFormatException
+import com.datadog.android.sessionreplay.internal.utils.MiscUtils
+import com.google.gson.JsonObject
+import okhttp3.MediaType.Companion.toMediaTypeOrNull
+import okhttp3.MultipartBody
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
+import kotlin.jvm.Throws
+
+internal class ResourceRequestBodyFactory(
+    private val internalLogger: InternalLogger
+) {
+
+    @Throws(InvalidPayloadFormatException::class)
+    internal fun create(
+        resources: List<RawBatchEvent>
+    ): RequestBody {
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
+        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
+        val applicationId = getApplicationId(resources)
+
+        resources.filter {
+            val metadata: JsonObject? = MiscUtils.safeDeserializeToJsonObject(it.metadata)
+            val elementApplicationId: String = metadata?.get(APPLICATION_ID_KEY)?.asString ?: ""
+            elementApplicationId == applicationId
+        }
+
+        addResourcesSection(builder, resources)
+        addApplicationIdSection(builder, applicationId)
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
+        return builder.build()
+    }
+
+    private fun getApplicationId(resources: List<RawBatchEvent>): String {
+        val resourcesMetaData: List<JsonObject> = resources.mapNotNull {
+            MiscUtils.safeDeserializeToJsonObject(it.metadata)
+        }
+
+        val applicationIds = resourcesMetaData.mapNotNull {
+            it.get(APPLICATION_ID_KEY)
+        }
+
+        if (applicationIds.isEmpty()) {
+            @Suppress("ThrowingInternalException")
+            throw InvalidPayloadFormatException(UNABLE_GET_APPLICATION_ID_ERROR)
+        }
+
+        var selectedApplicationId = applicationIds[0].asString
+
+        if (applicationIds.size > 1) {
+            internalLogger.log(
+                level = InternalLogger.Level.ERROR,
+                target = InternalLogger.Target.USER,
+                { MULTIPLE_APPLICATION_ID_ERROR }
+            )
+
+            var maxSize = 0
+            val groupedByApplicationId = resourcesMetaData.groupBy {
+                it.get(APPLICATION_ID_KEY)
+            }
+
+            groupedByApplicationId.forEach {
+                if (it.value.size > maxSize) {
+                    maxSize = it.value.size
+                    selectedApplicationId = it.key.asString
+                }
+            }
```

Comment:
Can `application_id` be changed during the SDK run? I doubt this can be the case, unless RUM feature is stopped and another one is registered, but we don't have such possibility right now.